File: software-cracking-10weeks/code/week10/serial_check.py
# -*- coding: utf-8 -*-
"""
10_serial_check.py — Thuật toán kiểm tra serial + keygen (minh hoạ, cho app của bạn).

Mục tiêu: cho thấy vì sao kiểm tra serial client-side dễ bị reverse; từ đó
          hiểu cách bảo vệ (chuyển kiểm tra lên server, dùng chữ ký số).
Đầu vào : --check <serial> (kiểm tra) hoặc --keygen <name> (sinh serial).
Đầu ra : kết quả kiểm tra hoặc serial sinh ra.
An toàn : chỉ minh hoạ thuật toán trên dữ liệu giả, không liên quan phần mềm thật.
"""
import argparse
import hashlib


def derive_serial(name: str) -> str:
    """Sinh serial theo tên (thuật toán đơn giản chỉ để minh hoạ)."""
    h = hashlib.sha256(("RE-" + name.upper()).encode()).hexdigest()
    return "RE-" + h[:8].upper()


def check_serial(name: str, serial: str) -> bool:
    """Kiểm tra serial có khớp tên hay không."""
    return serial.upper() == derive_serial(name)


def main() -> None:
    p = argparse.ArgumentParser(description="Minh hoạ kiểm tra serial + keygen.")
    p.add_argument("--check", nargs=2, metavar=("NAME", "SERIAL"), help="Kiểm tra serial.")
    p.add_argument("--keygen", metavar="NAME", help="Sinh serial hợp lệ cho tên.")
    args = p.parse_args()

    if args.keygen:
        print(f"Name   : {args.keygen}")
        print(f"Serial : {derive_serial(args.keygen)}")
    elif args.check:
        name, serial = args.check
        ok = check_serial(name, serial)
        print(f"Name   : {name}")
        print(f"Serial : {serial}")
        print(f"Kết quả: {'HỢP LỆ' if ok else 'KHÔNG hợp lệ'}")
    else:
        p.print_help()

    print("\nBài học: thuật toán nằm ngay trong client thì ai cũng reverse được.")
    print("Phòng thủ: kiểm tra phía server, ký số serial, chống debug (anti-tamper).")


if __name__ == "__main__":
    main()

File: software-cracking-10weeks/code/week10/serial_check.cpp
// 10_serial_check.cpp — Thuật toán kiểm tra serial + keygen (minh hoạ)
// Mục tiêu: cho thấy vì sao kiểm tra serial client-side dễ bị reverse; từ đó
//           hiểu cách bảo vệ (server-side, chữ ký số).
// An toàn: thuần tính toán trên chuỗi giả, không liên quan phần mềm thật.  g++ -std=c++17 -O2 serial_check.cpp -o demo
#include <iostream>
#include <string>
#include <cctype>

// "Hash" đơn giản: tổng trọng số các ký tự (minh hoạ, không phải hàm băm an toàn).
unsigned hash_name(const std::string& name) {
    unsigned h = 0x811C9DC5u;
    for (char c : name) { h ^= (unsigned char)std::toupper(c); h *= 0x01000193u; }
    return h;
}
std::string derive_serial(const std::string& name) {
    char buf[16]; std::snprintf(buf, sizeof(buf), "RE-%08X", hash_name("RE-" + name));
    return buf;
}
bool check(const std::string& name, const std::string& serial) {
    std::string s = serial; for (auto& c : s) c = std::toupper(c);
    return s == derive_serial(name);
}

int main(int argc, char** argv) {
    if (argc == 3 && std::string(argv[1]) == "--check") {
        std::cout << "Name: " << argv[2] << "\nSerial: " << derive_serial(argv[2]) << "\n";
    } else if (argc == 4 && std::string(argv[1]) == "--verify") {
        bool ok = check(argv[2], argv[3]);
        std::cout << "Name: " << argv[2] << "\nSerial: " << argv[3] << "\nKết quả: " << (ok ? "HỢP LỆ" : "KHÔNG hợp lệ") << "\n";
    } else {
        std::cout << "Dùng: ./demo --check <name>  |  ./demo --verify <name> <serial>\n";
        return 0;
    }
    std::cout << "\nBài học: thuật toán nằm trong client thì ai cũng reverse được.\n";
    std::cout << "Phòng thủ: kiểm tra phía server, ký số serial, chống debug.\n";
    return 0;
}
